### ISSUE TITLE
Slow on Llama_cpp_python

### ISSUE THREAD

**Issue by dnhkng** [kddubey/cappr#4]
This seems to work nicely for my use case, but using the llama_cpp_python backend with a long prompt and a dozen completions takes about 1 minute. Any tips on improving speed?

**Comment by kddubey** [comment]
Hi @dnhkng,

Unfortunately `cappr.llama_cpp` can be slow. This is partly b/c there currently isn't a way to batch over completions like we can in `cappr.huggingface`. The CAPPr algorithm also doesn't scale well, as discussed [here](https://cappr.readthedocs.io/en/latest/computational_performance.html#weaknesses).

2 different routes to improve performance:

1. Shorten the prompt. Perhaps if you post the prompt and completions we can figure out ways to make it shorter.

2. Don't use CAPPr. There are [other LLM structuring tools](https://www.reddit.com/r/LocalLLaMA/comments/17a4zlf/reliable_ways_to_get_structured_output_from_llms/) which provide a "just pick" functionality. I'm most familiar with the algorithm for [`guidance.select`](https://github.com/guidance-ai/guidance?tab=readme-ov-file#select-basic), and it's quite efficient.

Apologies for the slowness, and thanks for raising the issue

**Comment by dnhkng** [comment]
I do want to use CAPPr. 

I need the probabilities (yes I know that we can't use these as confidence values), but I think I have found something useful. 

I need to run a few thousand inference over a long prompt, with a large number of categories to test this though.

Any other tips would be welcome.

**Comment by kddubey** [comment]
1. Are you using a GPU? If not (and assuming it's ok to use a cloud GPU for security/compliance), can you point to the GGUF model you're using? Maybe the model and task will fit on a free or cheap GPU

2. Can you use a smaller version of the model? Also, if there's a PyTorch/AutoGPTQ version of the model, `cappr.huggingface` may be faster. Lmk which model you're using

3. Can you post the prompt and completions (or something similar-enough to work with)? In case you're doing in-context learning / few-shot prompting, consider reducing or eliminating the examples / do a zero-shot prompt. Sometimes, the lift from in-context learning is marginal

**Comment by kddubey** [comment]
Another tip: if all of the prompts start with the same long substring, e.g., system instructions and few-shot examples, use [`cappr.llama_cpp.cache`](https://cappr.readthedocs.io/en/latest/cappr.llama_cpp.classify.html#cappr.llama_cpp.classify.cache)

**Comment by dnhkng** [comment]
Thanks for the tips!
1) yes, using a beefy GPU
2) A big model is required, unfortunately.
3) the prompt is necessarily long, no way around that. 

I'm currently using [Outlines](https://github.com/outlines-dev/outlines) for category selection, but I want to try a new approach that requires probabilities. It initially, I planned on doing the naive 'tokenise the categories', and collect the probabilities for each token, and do the whole thing in a batch the size of the categories.  

But I assume that sometimes the same category wording can be achieved via multiple token combinations... So I found your library to use instead.

**Comment by kddubey** [comment]
Makes sense. How long are the completions? Despite the long prompt, it's still surprising to me that 12 completions takes a minute. Maybe the prompt can be refactored a bit to work with [`cappr.llama_cpp.cache_model`](https://cappr.readthedocs.io/en/latest/cappr.llama_cpp.classify.html#cappr.llama_cpp.classify.cache_model).  And is there an AutoGPTQ version of the big model?

Also, I found this point from your previous comment interesting:

> I need the probabilities (yes I know that we can't use these as confidence values), but I think I have found something useful.

Can you elaborate on this / point to a reference? I've [barely studied](https://cappr.readthedocs.io/en/latest/statistical_performance.html#calibration) this

**Comment by dnhkng** [comment]
I just mean the fact that LLMs are trained on cross entropy loss on next tokens. This leads to overconfidence, as the loss does not penalise miscalibration.  But maybe that's not a huge issue?

**Comment by kddubey** [comment]
> This leads to overconfidence, as the loss does not penalise miscalibration.

I believe that research is about neural networks leading to overconfidence. Cross entropy (CE) / negative log-likelihood (NLL) is also used to fit logistic regression, for example. A short mathematical argument will make it clear that CE/NLL is great for calibration: the loss is $-\log \hat{y}$, where $\hat{y}$ is the predicted probability of the given class. As $\hat{y}$ gets closer to 0 / further from 1, the loss goes up. In fact, it's unbounded. So CE/NLL penalizes miscalibrated models much more than other losses like mean squared error.

That aside, I take your point. LLMs are NNs, and they shouldn't be expected to be calibrated. Though I think it's worth researching how calibrated CAPPr's probabilities are. Hopefully they turn out to be helpful for your task

**Comment by dnhkng** [comment]
Exactly.  The tests I'm planning are to see how it works in practice.

**Comment by kddubey** [comment]
> As far as the algorithm works, does it sum the probabilities of all paths?

No. It tokenizes `end_of_prompt + completion` and calculates the log-probability of each token given `prompt`. See [my post about it here](https://stats.stackexchange.com/q/601159/337906) and [this page of the documentation](https://cappr.readthedocs.io/en/latest/related_work.html)
